harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,412 @@
+import abc
+import logging
+import queue
+import threading
+import time
+from datetime import datetime, timezone
+from typing import Any, Dict, List
+
+import psutil
+
+from determined import core
+from determined.common import api, constants
+from determined.common.api import bindings
+
+try:
+    import pynvml
+except ImportError:
+    pynvml = None
+
+logger = logging.getLogger("determined.core")
+
+
+class ProfilerContext:
+    """
+    ``ProfilerContext`` gives access to the system profiling feature within Determined.
+
+    It is responsible for collecting system metrics at specified time intervals and reporting
+    them to the master.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        distributed: core.DistributedContext,
+    ) -> None:
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+        self._distributed = distributed
+        self._on = False
+
+        self._collector = None
+        self._shipper = None
+
+    def on(self, interval: int = 1) -> None:
+        """
+        Turns system profiling functionality on.
+
+        This method creates two threads, one that collects system metrics at specified time
+        intervals, and another that ships them to the master.
+
+        .. note::
+
+            This method is idempotent: if profiling is already on, this method is effectively a
+            no-op.
+
+        Arguments:
+            interval: time (in seconds) between each metric collection.
+
+        """
+        if self._on:
+            return
+
+        # Only enable on local chief workers.
+        if self._distributed.local_rank != 0:
+            return
+
+        logger.info("Starting system metrics profiling.")
+
+        metrics_queue = queue.Queue()
+        self._collector = _Collector(
```

Comment:
I like your metric-group class hierarchy, and both Collector and Metric look like good classes.

And though I need to read more, I also like how responsibilities are broken down between Collector and Shipper. This so far feels like a really nice design.